File: mysql-async/src/main/scala/com/github/mauricio/async/db/mysql/codec/SendLongDataEncoder.scala
package com.github.mauricio.async.db.mysql.codec

import java.nio.charset.Charset

import com.github.mauricio.async.db.mysql.binary.BinaryRowEncoder
import com.github.mauricio.async.db.mysql.message.client.{ClientMessage, SendLongDataMessage}
import com.github.mauricio.async.db.util.{Log, ByteBufferUtils}
import io.netty.buffer.{Unpooled, ByteBuf}
import io.netty.channel.ChannelHandlerContext
import io.netty.handler.codec.MessageToMessageEncoder

object SendLongDataEncoder {
  val log = Log.get[SendLongDataEncoder]
}

class SendLongDataEncoder(charset: Charset)
    extends MessageToMessageEncoder[SendLongDataMessage](classOf[SendLongDataMessage]) {

  import com.github.mauricio.async.db.mysql.codec.SendLongDataEncoder.log

  private final val rowEncoder = new BinaryRowEncoder(charset)

  def isLong(value: Any): Boolean = rowEncoder.encoderFor(value).isLong(value)

  def encode(ctx: ChannelHandlerContext, message: SendLongDataMessage, out: java.util.List[Object]): Unit = {
    val result: ByteBuf = encode(message)

    ByteBufferUtils.writePacketLength(result, 0)

    if ( log.isTraceEnabled ) {
      log.trace(s"Writing message ${message.toString}")
    }

    out.add(result)
  }

  private def encode(message: SendLongDataMessage): ByteBuf = {
    val buffer = ByteBufferUtils.packetBuffer()
    buffer.writeByte(ClientMessage.PreparedStatementSendLongData)
    buffer.writeBytes(message.statementId)
    buffer.writeShort(message.paramId)

    Unpooled.wrappedBuffer(buffer, encodeValue(message.value))
  }

  private def encodeValue(maybeValue: Any) : ByteBuf = {
    if ( maybeValue == null || maybeValue == None ) {
      throw new UnsupportedOperationException("Cannot encode NULL as long value")
    } else {
      val value = maybeValue match {
        case Some(v) => v
        case _ => maybeValue
      }
      val encoder = rowEncoder.encoderFor(value)
      encoder.encodeLong(value)
    }
  }

}

File: mysql-async/src/main/java/com/github/jasync/sql/db/mysql/codec/SendLongDataEncoder.kt
package com.github.jasync.sql.db.mysql.codec

import com.github.jasync.sql.db.mysql.message.client.ClientMessage
import com.github.jasync.sql.db.mysql.message.client.SendLongDataMessage
import com.github.jasync.sql.db.util.ByteBufferUtils
import io.netty.buffer.Unpooled
import io.netty.channel.ChannelHandlerContext
import io.netty.handler.codec.MessageToMessageEncoder
import mu.KotlinLogging


private val logger = KotlinLogging.logger {}

class SendLongDataEncoder
    : MessageToMessageEncoder<SendLongDataMessage>(SendLongDataMessage::class.java) {


    companion object {
        const val LONG_THRESHOLD = 1023
    }

    override fun encode(ctx: ChannelHandlerContext, message: SendLongDataMessage, out: MutableList<Any>) {
        logger.trace { "Writing message $message" }

        val sequence = 0

        val headerBuffer = ByteBufferUtils.mysqlBuffer(3 + 1 + 1 + 4 + 2)
        ByteBufferUtils.write3BytesInt(headerBuffer, 1 + 4 + 2 + message.value.readableBytes())
        headerBuffer.writeByte(sequence)

        headerBuffer.writeByte(ClientMessage.PreparedStatementSendLongData)
        headerBuffer.writeBytes(message.statementId)
        headerBuffer.writeShort(message.paramId)

        val result = Unpooled.wrappedBuffer(headerBuffer, message.value)

        out.add(result)
    }

}
